File: Britefury/DocModel/DMListCommandTracker.py
##-*************************
##-* This program is free software; you can use it, redistribute it and/or modify it
##-* under the terms of the GNU General Public License version 2 as published by the
##-* Free Software Foundation. The full text of the GNU General Public License
##-* version 2 can be found in the file named 'COPYING' that accompanies this
##-* program. This source code is (C)copyright Geoffrey French 1999-2007.
##-*************************
from Britefury.CommandHistory import CommandHistory
from Britefury.CommandHistory import CommandTracker




class DMListAppendCommand (CommandHistory.Command):
	def __init__(self, ls, x):
		super( DMListAppendCommand, self ).__init__()
		self._ls = ls
		self._x = x


	def execute(self):
		self._ls.append( self._x )

	def unexecute(self):
		del self._ls[-1]




class DMListExtendCommand (CommandHistory.Command):
	def __init__(self, ls, xs):
		super( DMListExtendCommand, self ).__init__()
		self._ls = ls
		self._xs = xs
		self._size = len( xs )


	def execute(self):
		self._ls.extend( self._xs )

	def unexecute(self):
		del self._ls[-self._size:]




class DMListInsertCommand (CommandHistory.Command):
	def __init__(self, ls, index, x):
		super( DMListInsertCommand, self ).__init__()
		self._ls = ls
		self._index = index
		self._x = x


	def execute(self):
		self._ls.insert( self._index, self._x )

	def unexecute(self):
		del self._ls[self._index]




class DMListRemoveCommand (CommandHistory.Command):
	def __init__(self, ls, x):
		super( DMListRemoveCommand, self ).__init__()
		self._ls = ls
		self._x = x
		self._index = ls.index( x )


	def execute(self):
		self._ls.remove( self._x )

	def unexecute(self):
		self._ls.insert( self._index, self._x )




class DMListSetCommand (CommandHistory.Command):
	def __init__(self, ls, oldContents, contents):
		super( DMListSetCommand, self ).__init__()
		self._ls = ls
		self._oldContents = oldContents
		self._contents = contents


	def execute(self):
		self._ls[:] = self._contents

	def unexecute(self):
		self._ls[:] = self._oldContents



	def canJoinWith(self, command):
		"Determine if @command can be joined with @self"
		return isinstance( command, DMListSetCommand )  and  self._ls is command._ls  and  not self._bFinished

	def joinWith(self, command):
		"Join @command with @self, where @command took place after @self, and the result should go into @self"
		self._contents = command._contents





class DMListCommandTracker (CommandTracker.CommandTracker):
	def track(self, ls):
		super( DMListCommandTracker, self ).track( ls )

		for x in ls:
			if self._commandHistory.canTrack( x ):
				self._commandHistory.track( x )


	def stopTracking(self, ls):
		for x in ls:
			if self._commandHistory.canTrack( x ):
				self._commandHistory.stopTracking( x )

		super( DMListCommandTracker, self ).stopTracking( ls )




	def _f_onAppended(self, ls, x):
		if self._commandHistory.canTrack( x ):
			self._commandHistory.track( x )
		self._commandHistory.addCommand( DMListAppendCommand( ls, x ) )


	def _f_onExtended(self, ls, xs):
		self._commandHistory.addCommand( DMListExtendCommand( ls, xs ) )
		for x in xs:
			if self._commandHistory.canTrack( x ):
				self._commandHistory.track( x )


	def _f_onInserted(self, ls, index, x):
		if self._commandHistory.canTrack( x ):
			self._commandHistory.track( x )
		self._commandHistory.addCommand( DMListInsertCommand( ls, index, x ) )


	def _f_onRemove(self, ls, x):
		self._commandHistory.addCommand( DMListRemoveCommand( ls, x ) )
		if self._commandHistory.canTrack( x ):
			self._commandHistory.stopTracking( x )


	def _f_onSet(self, ls, oldContents, contents):
		oldContentsSet = set( oldContents )
		contentsSet = set( contents )
		for x in oldContentsSet - contentsSet:
			if self._commandHistory.canTrack( x ):
				self._commandHistory.stopTracking( x )
		self._commandHistory.addCommand( DMListSetCommand( ls, oldContents, contents ) )
		for x in contentsSet - oldContentsSet:
			if self._commandHistory.canTrack( x ):
				self._commandHistory.track( x )

File: java/BritefuryJ/DocModel/DMListCommandTracker.java
//##* This program is free software; you can use it, redistribute it and/or modify it
//##* under the terms of the GNU General Public License version 2 as published by the
//##* Free Software Foundation. The full text of the GNU General Public License
//##* version 2 can be found in the file named 'COPYING' that accompanies this
//##* program. This source code is (C)copyright Geoffrey French 2008.
//##************************
package BritefuryJ.DocModel;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import BritefuryJ.CommandHistory.Command;
import BritefuryJ.CommandHistory.CommandHistory;
import BritefuryJ.CommandHistory.CommandTracker;
import BritefuryJ.CommandHistory.CommandTrackerFactory;
import BritefuryJ.CommandHistory.Trackable;

class DMListCommandTracker extends CommandTracker
{
	private static class AddCommand extends Command
	{
		private DMList xls;
		private Object x;
		
		public AddCommand(DMList ls, Object x)
		{
			this.xls = ls;
			this.x = x;
		}

		
		protected void execute()
		{
			xls.add( x );
		}

		protected void unexecute()
		{
			xls.remove( xls.size() - 1 );
		}

		protected String getDescription()
		{
			return "List add";
		}
	}

	
	
	private static class InsertCommand extends Command
	{
		private DMList ls;
		private int i;
		private Object x;
		
		public InsertCommand(DMList ls, int i, Object x)
		{
			this.ls = ls;
			this.i = i;
			this.x = x;
		}

		
		protected void execute()
		{
			ls.add( i, x );
		}

		protected void unexecute()
		{
			ls.remove( i );
		}

		protected String getDescription()
		{
			return "List insert (" + i + ")";
		}
	}

	
	
	private static class AddAllCommand extends Command
	{
		private DMList ls;
		private List<Object> x;
		
		public AddAllCommand(DMList ls, List<Object> x)
		{
			this.ls = ls;
			this.x = x;
		}

		
		protected void execute()
		{
			ls.addAll( x );
		}

		protected void unexecute()
		{
			ls.removeLast( x.size() );
		}

		protected String getDescription()
		{
			return "List add all";
		}
	}

	
	
	private static class InsertAllCommand extends Command
	{
		private DMList ls;
		private int i;
		private List<Object> x;
		
		public InsertAllCommand(DMList ls, int i, List<Object> x)
		{
			this.ls = ls;
			this.i = i;
			this.x = x;
		}

		
		protected void execute()
		{
			ls.addAll( i, x );
		}

		protected void unexecute()
		{
			ls.removeRange( i, x.size() );
		}

		protected String getDescription()
		{
			return "List insert all (" + i + ")";
		}
	}

	
	
	private static class ClearCommand extends Command
	{
		private DMList ls;
		private ArrayList<Object> contents;
		
		public ClearCommand(DMList ls, ArrayList<Object> contents)
		{
			this.ls = ls;
			this.contents = contents;
		}

		
		protected void execute()
		{
			ls.clear();
		}

		protected void unexecute()
		{
			ls.addAll( contents );
		}

		protected String getDescription()
		{
			return "List clear";
		}
	}

	
	private static class RemoveCommand extends Command
	{
		private DMList ls;
		private int i;
		private Object x;
		
		public RemoveCommand(DMList ls, int i, Object x)
		{
			this.ls = ls;
			this.i = i;
			this.x = x;
		}

		
		protected void execute()
		{
			ls.remove( i );
		}

		protected void unexecute()
		{
			ls.add( i, x );
		}

		protected String getDescription()
		{
			return "List remove (" + i + ")";
		}
	}


	
	private static class SetCommand extends Command
	{
		private DMList ls;
		private int i;
		private Object oldX, x;
		
		public SetCommand(DMList ls, int i, Object oldX, Object x)
		{
			this.ls = ls;
			this.i = i;
			this.oldX = oldX;
			this.x = x;
		}

		
		protected void execute()
		{
			ls.set( i, x );
		}

		protected void unexecute()
		{
			ls.set( i, oldX );
		}

		protected String getDescription()
		{
			return "List set (" + i + ")";
		}
	}

	
	
	private static class SetContentsCommand extends Command
	{
		private DMList ls;
		private Object[] oldContents;
		private Object[] newContents;
		
		public SetContentsCommand(DMList ls, Object[] oldContents, Object[] newContents)
		{
			this.ls = ls;
			this.oldContents = oldContents;
			this.newContents = newContents;
		}

		
		protected void execute()
		{
			ls.commandTracker_setContents( newContents );
		}

		protected void unexecute()
		{
			ls.commandTracker_setContents( oldContents );
		}

		protected String getDescription()
		{
			return "List set contents";
		}
	}

	private static class RemoveLastCommand extends Command
	{
		private DMList ls;
		private Object removedValues[];
		
		public RemoveLastCommand(DMList ls, Object removedValues[])
		{
			this.ls = ls;
			this.removedValues = removedValues;
		}

		
		protected void execute()
		{
			ls.removeLast( removedValues.length );
		}

		protected void unexecute()
		{
			ls.addAll( Arrays.asList( removedValues ) );
		}

		protected String getDescription()
		{
			return "List remove last";
		}
	}

	private static class RemoveRangeCommand extends Command
	{
		private DMList ls;
		private int pos;
		private Object removedValues[];
		
		public RemoveRangeCommand(DMList ls, int pos, Object removedValues[])
		{
			this.ls = ls;
			this.pos = pos;
			this.removedValues = removedValues;
		}

		
		protected void execute()
		{
			ls.removeRange( pos, removedValues.length );
		}

		protected void unexecute()
		{
			ls.addAll( pos, Arrays.asList( removedValues ) );
		}

		protected String getDescription()
		{
			return "List remove range";
		}
	}

	
	
	
	
	public DMListCommandTracker(CommandHistory commandHistory)
	{
		super( commandHistory );
	}
	
	
	
	protected void track(Trackable t)
	{
		super.track( t );
		
		DMList ls = (DMList)t;
		for (Object x: ls)
		{
			commandHistory.track( x );
		}
	}
	
	protected void stopTracking(Trackable t)
	{
		DMList ls = (DMList)t;
		for (Object x: ls)
		{
			commandHistory.stopTracking( x );
		}

		super.stopTracking( t );
	}
	
	
	
	protected void onAdd(DMList ls, Object x)
	{
		commandHistory.track( x );
		commandHistory.addCommand( new AddCommand( ls, x ) );
	}

	protected void onInsert(DMList ls, int i, Object x)
	{
		commandHistory.track( x );
		commandHistory.addCommand( new InsertCommand( ls, i, x ) );
	}

	protected void onAddAll(DMList ls, List<Object> xs)
	{
		for (Object x: xs)
		{
			commandHistory.track( x );
		}
		commandHistory.addCommand( new AddAllCommand( ls, xs ) );
	}

	protected void onInsertAll(DMList ls, int i, List<Object> xs)
	{
		for (Object x: xs)
		{
			commandHistory.track( x );
		}
		commandHistory.addCommand( new InsertAllCommand( ls, i, xs ) );
	}

	protected void onClear(DMList ls, ArrayList<Object> contents)
	{
		commandHistory.addCommand( new ClearCommand( ls, contents ) );
		for (Object x: contents)
		{
			commandHistory.stopTracking( x );
		}
	}

	protected void onRemove(DMList ls, int i, Object x)
	{
		commandHistory.addCommand( new RemoveCommand( ls, i, x ) );
		commandHistory.stopTracking( x );
	}

	protected void onSet(DMList ls, int i, Object oldX, Object x)
	{
		commandHistory.track( x );
		commandHistory.addCommand( new SetCommand( ls, i, oldX, x ) );
		commandHistory.stopTracking( oldX );
	}

	protected void onSetContents(DMList ls, Object[] oldContents, Object[] newContents)
	{
		for (Object x: newContents)
		{
			commandHistory.track( x );
		}
		commandHistory.addCommand( new SetContentsCommand( ls, oldContents, newContents ) );
		for (Object oldX: oldContents)
		{
			commandHistory.stopTracking( oldX );
		}
	}
	
	protected void onRemoveLast(DMList ls, Object[] removedValues)
	{
		commandHistory.addCommand( new RemoveLastCommand( ls, removedValues ) );
		for (Object x: removedValues)
		{
			commandHistory.stopTracking( x );
		}
	}
	
	protected void onRemoveRange(DMList ls, int pos, Object[] removedValues)
	{
		commandHistory.addCommand( new RemoveRangeCommand( ls, pos, removedValues ) );
		for (Object x: removedValues)
		{
			commandHistory.stopTracking( x );
		}
	}
	
	
	public static CommandTrackerFactory factory = new CommandTrackerFactory()
	{
		@Override
		public CommandTracker createTracker(CommandHistory history)
		{
			return new DMListCommandTracker( history );
		}
	};
}
